In [2]:
pip install plotly


[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
# Importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go

orignal_data = pd.read_csv('GRAND_FINALE.csv')
orignal_data.head()

PS ID       Organization  S No  Idea ID  Team ID            Team Name  \
0  SIH1524  Godrej Appliances     1    54238    47431  The Dynamic Coolers   
1  SIH1524  Godrej Appliances     2    30628    25943        NeuralNexus25   
2  SIH1524  Godrej Appliances     3    19233    18051            Arka_Team   
3  SIH1524  Godrej Appliances     4    25132    36557            MECHSPACE   
4  SIH1524  Godrej Appliances     5    26721    20483           Pure Rince   

   Team Leader Name AISHE Code  \
0  M Priyadharshini     U-0461   
1    Joel Stelton M    C-16547   
2     Meeraa Chadha    C-47474   
3            Gokuls    C-37058   
4       Nishit Modi     U-0135   

                               Name of the Institute         City  \
0  M.G.R. Educational and Research Institute, Che...  Thiruvallur   
1                     Sri Sairam Engineering College  Kanchipuram   
2   NATIONAL INSTITUTE OF FASHION TECHNOLOGY, KANGRA       Kangra   
3                  Knowledge Institute of Technology        Salem   
4        Gujarat Technological University, Ahmedabad    Ahmadabad   

              State    Status  
0        Tamil Nadu  SELECTED  
1        Tamil Nadu  SELECTED  
2  Himachal Pradesh  SELECTED  
3        Tamil Nadu  SELECTED  
4           Gujarat  SELECTED

In [4]:
# Knowing About the Columns
orignal_data.columns

Index(['PS ID', 'Organization', 'S No', 'Idea ID', 'Team ID', 'Team Name',
       'Team Leader Name', 'AISHE Code', 'Name of the Institute', 'City',
       'State', 'Status'],
      dtype='object')

In [5]:
# Removing the unnecessary columns
data =orignal_data.drop(columns=[ 'S No','Idea ID', 'Name of the Institute','Team Leader Name','Team ID'], axis=1)
data.head(10)

PS ID       Organization            Team Name AISHE Code  \
0  SIH1524  Godrej Appliances  The Dynamic Coolers     U-0461   
1  SIH1524  Godrej Appliances        NeuralNexus25    C-16547   
2  SIH1524  Godrej Appliances            Arka_Team    C-47474   
3  SIH1524  Godrej Appliances            MECHSPACE    C-37058   
4  SIH1524  Godrej Appliances           Pure Rince     U-0135   
5  SIH1525  Godrej Appliances         Refrriranger    C-28993   
6  SIH1525  Godrej Appliances            <Archemy>    C-26795   
7  SIH1525  Godrej Appliances             Cerberus     U-0760   
8  SIH1525  Godrej Appliances        Idea Infinity    C-49143   
9  SIH1525  Godrej Appliances         KALKI CODERS    C-29461   

                  City             State    Status  
0          Thiruvallur        Tamil Nadu  SELECTED  
1          Kanchipuram        Tamil Nadu  SELECTED  
2               Kangra  Himachal Pradesh  SELECTED  
3                Salem        Tamil Nadu  SELECTED  
4            Ahmadabad           Gujarat  SELECTED  
5  Gautam Buddha Nagar     Uttar Pradesh  SELECTED  
6       Ramanathapuram        Tamil Nadu  SELECTED  
7             Chittoor    Andhra Pradesh  SELECTED  
8         Virudhunagar        Tamil Nadu  SELECTED  
9          S.A.S Nagar            Punjab  SELECTED

Some columns are not required such as Team Leader Name and Name of the Institute they can be reffered through the unique Team Id and AISHE Cdoe for institute

In [6]:
data.groupby("State").size().reset_index(name="Number of Selected Teams")

State  Number of Selected Teams
0      Andhra Pradesh                        61
1               Assam                         5
2               Bihar                        14
3          Chandigarh                         7
4        Chhattisgarh                        22
5               Delhi                        50
6             Gujarat                        56
7             Haryana                        21
8    Himachal Pradesh                         3
9   Jammu and Kashmir                         6
10          Jharkhand                        12
11          Karnataka                        63
12             Kerala                         7
13     Madhya Pradesh                        76
14        Maharashtra                       303
15            Manipur                         4
16             Odisha                        26
17         Puducherry                         6
18             Punjab                        29
19          Rajasthan                        33
20             Sikkim                         1
21         TAMIL NADU                         1
22         Tamil Nadu                       315
23          Telangana                        50
24            Tripura                         1
25      Uttar Pradesh                        87
26        Uttarakhand                         8
27        West Bengal                        72

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1339 entries, 0 to 1338
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   PS ID         1339 non-null   object
 1   Organization  1339 non-null   object
 2   Team Name     1339 non-null   object
 3   AISHE Code    1339 non-null   object
 4   City          1339 non-null   object
 5   State         1339 non-null   object
 6   Status        1339 non-null   object
dtypes: object(7)
memory usage: 73.4+ KB


Observation:
1) Total teams selected for grand finale are 1339 across India.

In [8]:
data.isnull().sum()

PS ID           0
Organization    0
Team Name       0
AISHE Code      0
City            0
State           0
Status          0
dtype: int64

Observation:<br>
2) No missing values are present in any column of the dataset.

In [9]:
#Changing the name from TAMIL NADU TO Tamil Nadu
data[data["State"]=="TAMIL NADU"]
data.loc[data['State'] == "TAMIL NADU", 'State'] = 'Tamil Nadu'

In [10]:
print(data.describe(include='all'))

          PS ID                   Organization    Team Name AISHE Code  City  \
count      1339                           1339         1339       1339  1339   
unique      254                             45         1339        650   195   
top     SIH1572  AICTE, MIC-Student Innovation  LAWGORITHM_    C-16547  Pune   
freq         11                            182            1         16    95   

             State    Status  
count         1339      1339  
unique          27         2  
top     Tamil Nadu  SELECTED  
freq           316      1324  


Observation:<br>
3) Total Number of Unique PS ID are:254.<br>
4) Total Number of Unique Orgaization are:45.<br>
5) Total Number of Unique AISHE Code(Colleges) are:650.<br>
6) Total Number of Unique City are:195.<br>
7) Total Number of Unique State are:28.




Correcting the Error:<br>
Tamil Nadu State has been marked two timmes in dataset using different names. Changing both the names to a sinngle name


In [11]:
print("TOTAL NUMBER OF ORGANIZATIONS:45")
print("Organizations with Maximum submitted ideas:",data['Organization'].value_counts().index[0])
print("Organizations with Minimum submitted ideas:",data['Organization'].value_counts().index[-1])


TOTAL NUMBER OF ORGANIZATIONS:45
Organizations with Maximum submitted ideas: AICTE, MIC-Student Innovation
Organizations with Minimum submitted ideas: Ministry of Consumer Affairs, Food and Public Distribution


In [12]:
print("PS ID With:")
print("Maximum Selection is",data['PS ID'].value_counts().index[0],"in which",
      data['PS ID'].value_counts().unique()[0],"Teams are selected.")

print("Minimum Selection is",data['PS ID'].value_counts().index[-1],"in which",
      data['PS ID'].value_counts().unique()[-1],"Teams are selected.")

PS ID With:
Maximum Selection is SIH1572 in which 11 Teams are selected.
Minimum Selection is SIH1737 in which 3 Teams are selected.


In [13]:
print("State With:")
print("Maximum Selection is",data['State'].value_counts().index[0],"in which",
      data['State'].value_counts().values[0],"Teams are selected.")

print("Minimum Selection are",data['State'].value_counts().index[-1],"and",data['State'].value_counts().index[-2],"in which",
      data['State'].value_counts().values[-1],"Team is selected.")

State With:
Maximum Selection is Tamil Nadu in which 316 Teams are selected.
Minimum Selection are Sikkim and Tripura in which 1 Team is selected.


In [14]:
df=data.groupby("State").size().reset_index(name="Number Of Teams Selected")
df.sort_values(by='Number Of Teams Selected')

State  Number Of Teams Selected
20             Sikkim                         1
23            Tripura                         1
8    Himachal Pradesh                         3
15            Manipur                         4
1               Assam                         5
17         Puducherry                         6
9   Jammu and Kashmir                         6
12             Kerala                         7
3          Chandigarh                         7
25        Uttarakhand                         8
10          Jharkhand                        12
2               Bihar                        14
7             Haryana                        21
4        Chhattisgarh                        22
16             Odisha                        26
18             Punjab                        29
19          Rajasthan                        33
5               Delhi                        50
22          Telangana                        50
6             Gujarat                        56
0      Andhra Pradesh                        61
11          Karnataka                        63
26        West Bengal                        72
13     Madhya Pradesh                        76
24      Uttar Pradesh                        87
14        Maharashtra                       303
21         Tamil Nadu                       316

In [15]:
df1=data.groupby("Organization").size().reset_index(name="Number Of Teams Selected")
df1=df1.sort_values(by="Number Of Teams Selected",ascending=False).reset_index()
df1.drop(columns="index",axis=1)

Organization  \
0                       AICTE, MIC-Student Innovation   
1                              Ministry of Jal Shakti   
2                          Government of NCT of Delhi   
3     National Technical Research Organisation (NTRO)   
4                           Ministry of Communication   
5          Ministry of Social Justice and Empowerment   
6                                   Ministry of Power   
7           Indian Space Research Organization (ISRO)   
8                               Madhya Pradesh Police   
9               Ministry of Housing and Urban Affairs   
10               Department of Science and Technology   
11  Ministry of Electronics and Information Techno...   
12                                Ministry of Defence   
13                         Ministry of Earth Sciences   
14  National Critical Information Infrastructure P...   
15        Ministry of Agriculture and Farmers Welfare   
16  Ministry of Skill Development and Entrepreneur...   
17                              Ministry of Education   
18                        Ministry of Law and Justice   
19                                   Ministry of Coal   
20                                  Ministry of AYUSH   
21                 Ministry of Youth Affairs & Sports   
22                               Government of Punjab   
23             Indo-Tibetan Border Police (ITBP), MHA   
24                                              AICTE   
25        GAIL, Ministry of Petroleum and Natural Gas   
26                                           Autodesk   
27                   Bharat Electronics Limited (BEL)   
28                            Government of Rajasthan   
29                     Narcotics Control Bureau (NCB)   
30  Ministry of Fisheries, Animal Husbandry and Da...   
31                              Government of Gujarat   
32                                Ministry of Railway   
33                National Investigation Agency (NIA)   
34   The National Disaster Response Force (NDRF), MHA   
35                                  Godrej Appliances   
36                 National Security Guard (NSG), MHA   
37                                Ministry of Culture   
38                         Ministry of Panchayati Raj   
39                    MathWorks India Private Limited   
40                                              AYUSH   
41                     Government of Himachal Pardesh   
42  Ministry of Consumer Affairs, Food and Public ...   
43           Ministry of Information and Broadcasting   
44         National Aluminium Company Limited (NALCO)   

    Number Of Teams Selected  
0                        182  
1                        114  
2                         84  
3                         70  
4                         63  
5                         60  
6                         55  
7                         48  
8                         41  
9                         41  
10                        37  
11                        32  
12                        32  
13                        31  
14                        30  
15                        29  
16                        26  
17                        25  
18                        25  
19                        21  
20                        20  
21                        17  
22                        16  
23                        16  
24                        15  
25                        15  
26                        15  
27                        15  
28                        15  
29                        15  
30                        15  
31                        15  
32                        11  
33                        11  
34                        10  
35                        10  
36                        10  
37                        10  
38                        10  
39                         6  
40                         6  
41                         5  
42                         5  
43                         5  
44                     

# Visualization through different Graphs 

In [16]:
df = data[data['Organization'] == "AICTE, MIC-Student Innovation"]
ps_id_counts = df['PS ID'].value_counts()

fig = px.bar(x=ps_id_counts.index, y=ps_id_counts.values, text_auto='.2s',title="PS ID of Organization(AICTE)")

fig.update_traces(
    textfont_size=12, 
    textposition="outside", 
    cliponaxis=False,
    marker_color='royalblue', 
    textfont_color='white')

fig.update_layout(
    xaxis_title="PS ID(With more than 1 Team Selected)", 
    yaxis_title="Number of Teams Selected", 
    title_font_size=18,
    title_font_family='Arial',
    plot_bgcolor='rgba(240, 240, 240, 1)')
fig.show()


Observation:<br>
8) Maximum Number of team Selected for a PS ID is 8.<br>
9) There are total 9 PS ID's with 6 teams selected and 24 PS ID's with 5 teams selected .<br>





In [17]:
df=data["State"].value_counts()
greaterthan_50=df[df>=50]
lessthan_50=df[df<50]

fig = px.bar(df, x = greaterthan_50.index, y = greaterthan_50.values, color = [14, 12, 8, 10, 16,15,19,20,21,22],)
fig.update_layout(
    title='STATES(With more than 50 selections)',
    xaxis_title='Name of States',
    yaxis_title='Number of Teams Selected',
    xaxis_tickangle=-45 )
fig.show()


Observation:<br>
10) There are Total of 10 States in which more than 50 Teams are selected.<br>
11) Maximum Number of Teams are Selected from Tamil Nadu.<br>
12) TOP 5 STATES ARE:<br>
    Tamil Nadu<br>
    Maharashtra<br>
    Uttar pradesh<br>
    Madhya Pradesh<br>
    West Bengal




In [18]:
fig = px.bar(df,
    x=lessthan_50.index,
    y=lessthan_50.values,
    color = [14, 12, 8, 10, 16,15,19,20,21,22,25,28,29,35,38,45,50]
)
fig.update_layout(
    title='STATES(With less than 50 selections)',
    xaxis_title='Name of States',
    yaxis_title='Number of Teams Selected',
    xaxis_tickangle=-45 )
fig.show()

Observation:<br>
13) There are Total of 17 States in which less than 50 Teams are selected.<br>
14) Minimum Number of Teams are Selected from Tripura and Sikkim.<br>

In [19]:
df=data
df=df['City'].value_counts()
g30=df[df>30]
print("Cities With more than 30 Selections")
print(g30.reset_index())
l30=df[df<30]
print("Cities With less than 30 Selections")
print(l30.reset_index())
print("Cities With only 1 Team Selected")
a=df[df==1]
print(a.reset_index())

Cities With more than 30 Selections
              City  count
0             Pune     95
1       Coimbatore     81
2      Kanchipuram     71
3  Mumbai Suburban     51
4           Mumbai     44
5  Bengaluru Urban     39
6      Thiruvallur     35
Cities With less than 30 Selections
                   City  count
0    24 Paraganas North     29
1                Indore     29
2                 Erode     27
3                Bhopal     25
4             Hyderabad     25
..                  ...    ...
183              Nirmal      1
184              Howrah      1
185             Dhanbad      1
186           Ernakulam      1
187             Hooghly      1

[188 rows x 2 columns]
Cities With only 1 Team Selected
                       City  count
0                    Kangra      1
1                 Sambalpur      1
2                Sangareddy      1
3              West Tripura      1
4              Mahendragarh      1
5                  Karaikal      1
6                     Samba      1
7          

In [20]:

city_counts = data.groupby(['State', 'City']).size().reset_index(name='Count')
top_states = city_counts.groupby('State')['Count'].sum().nlargest(5).index
filtered_data = city_counts[city_counts['State'].isin(top_states)]

top_cities_per_state = (
    filtered_data.groupby('State')
    .apply(lambda x: x.nlargest(3, 'Count'))
    .reset_index(drop=True))

fig = px.bar(
    top_cities_per_state, 
    x='City', 
    y='Count', 
    color='State',
    text='Count',
    title='Top 3 Cities for 5 States',
    labels={'Count': 'Number of Teams Selected', 'City': 'Name Of Cities'},
    hover_data=['State'])
fig.update_traces(
    marker_line_width=1.5, 
    marker_line_color='black', 
    textposition='outside'
)
fig.update_layout(
    barmode='group',
    xaxis_tickangle=-45,
    title_x=0.5
)

fig.show()

C:\Users\Tanishq Bhakar\AppData\Local\Temp\ipykernel_10732\773243406.py:7: DeprecationWarning:

DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.



OBSERVATIONS:<br>
15) Pune City in Maharashtra has maximum number of selections.<br>
16) Coimbatore City of Tamil Nadu has maximum of 81 selections.<br>
17) Indore City of Madhya Pradesh has maximum of 29 selections. <br> 
18) Ghaziabad City of Uttar Pradesh has maximum of 21 selections. <br> 
19) 24 Praganas North City of West Bengal has maximum of 29 selections.  


In [21]:
state_counts = data['State'].value_counts().reset_index()
state_counts.columns = ['State', 'Count']
above_threshold = state_counts[state_counts['Count'] > 10]
below_threshold = state_counts[state_counts['Count'] <= 10]

other_states = pd.DataFrame({'State': ['Other States'], 'Count': [below_threshold['Count'].sum()]})
final_data = pd.concat([above_threshold, other_states])
fig = px.pie(
    final_data, 
    names='State', 
    values='Count', 
    title="DISTRIBUTION OF STATES(on the basis of teams selected)",
    color_discrete_sequence=px.colors.qualitative.Set3 
)
fig.update_traces(textposition='inside', textinfo='percent+label')
fig.show()

In [22]:
data["Organization"].value_counts().values
org_counts = data['Organization'].value_counts().reset_index()
org_counts.columns = ['Organization', 'Count']
above_threshold = org_counts[org_counts['Count'] >= 10]
below_threshold = org_counts[org_counts['Count'] < 10]
other_orgs = pd.DataFrame({'Organization': ['Other Organizations'], 'Count': [below_threshold['Count'].sum()]})
final_data = pd.concat([above_threshold, other_orgs])

fig = px.pie(
    final_data, 
    names='Organization', 
    values='Count', 
    title="DISTRIBUTION OF ORGANIZATION(on the basis of teams selected)",
    color_discrete_sequence=px.colors.qualitative.Set3
)
fig.update_traces(textposition='inside', textinfo='percent+label')
fig.show()

In [23]:
import pandas as pd
import plotly.express as px

# Prepare data
org_counts = data['Organization'].value_counts().reset_index()
org_counts.columns = ['Organization', 'Count']
above_threshold = org_counts[org_counts['Count'] >= 10]
below_threshold = org_counts[org_counts['Count'] < 10]
other_orgs = pd.DataFrame({'Organization': ['Other Organizations'], 'Count': [below_threshold['Count'].sum()]})
final_data = pd.concat([above_threshold, other_orgs])

# Define a custom color palette
custom_colors = px.colors.qualitative.Set2

# Create the pie chart
fig = px.pie(
    final_data,
    names='Organization',
    values='Count',
    title="Distribution of Organizations (Based on Selected Teams)",
    color_discrete_sequence=custom_colors
)

# Update trace for better visuals
fig.update_traces(
    textposition='inside',
    textinfo='percent+label',
    pull=[0.1 if org == 'Other Organizations' else 0 for org in final_data['Organization']],  # Highlight "Other Organizations"
    hovertemplate='<b>%{label}</b><br>Count: %{value}<br>Percentage: %{percent}<extra></extra>'
)

# Customize layout
fig.update_layout(
    title=dict(
        text="Distribution of Organizations (Based on Selected Teams)",
        font=dict(size=20, color='darkblue'),
        x=0.5
    ),
    legend=dict(
        title="Organizations",
        font=dict(size=12),
        orientation='h',
        x=0.5,
        xanchor='center',
        y=-0.1
    )
)

# Show the chart
fig.show()


In [24]:
college_counts = orignal_data.groupby(['AISHE Code', 'Name of the Institute']).size().reset_index(name='No. Of Teams Selected')
top_10_colleges = college_counts.nlargest(10, 'No. Of Teams Selected')

fig = px.bar(
    top_10_colleges, 
    x='Name of the Institute', 
    y='No. Of Teams Selected', 
    text='No. Of Teams Selected', 
    title='Top 10 Colleges with Maximum Selections',
    labels={'Name of the Institute': 'College Name', 'No. Of Teams Selected': 'Number of Selections'},
    color='No. Of Teams Selected',  
    color_continuous_scale='Viridis'
)
fig.update_traces(
    textposition='outside', 
    marker_line_width=1.5, 
    marker_line_color='black'
)
fig.update_layout(
    xaxis_tickangle=-45,            
    title_x=0.5,                    
    font=dict(size=12),              
    xaxis=dict(
        tickfont=dict(size=10)       
    ),
    height=500,                      
    width=800                        
)
fig.show()
top_10_colleges

AISHE Code                              Name of the Institute  \
61     C-16547                     Sri Sairam Engineering College   
279    C-36995  Sri Krishna College of Engineering and Technology   
67     C-16576                    Chennai Institute of Technology   
76     C-16626                    Rajalakshmi Engineering College   
276    C-36984              Bannari Amman Institute of Technology   
296    C-37065                          Kongu Engineering College   
568     U-0535  Rajiv Gandhi Institute of Petroleum Technology...   
43     C-15788          WALCHAND INSTITUTE OF TECHNOLOGY, SOLAPUR   
66     C-16572            Sri Venkateswara College of Engineering   
175    C-27058                  Mepco Schlenk Engineering College   

     No. Of Teams Selected  
61                      16  
279                     15  
67                      14  
76                      11  
276                     11  
296                     11  
568                     10  
43                       9  
66                       9  
175                      9

OBSERVATIONS:<br>
20) Sri Sairam Engineering College has maximum number of 16 Teams selected from a single college.<br>


In [26]:
import plotly.express as px

# Top 10 Colleges Bar Chart
fig = px.bar(
    top_10_colleges, 
    x='Name of the Institute', 
    y='No. Of Teams Selected', 
    text='No. Of Teams Selected', 
    title='Top 10 Colleges with Maximum Selections',
    labels={'Name of the Institute': 'College Name', 'No. Of Teams Selected': 'Number of Selections'},
    color='No. Of Teams Selected',  
    color_continuous_scale=px.colors.sequential.Mint  # Use the Mint color palette
)

# Update traces for better visuals
fig.update_traces(
    textposition='outside',          # Place text above bars
    marker_line_width=1.5,           # Add border to bars
    marker_line_color='black'        # Border color
)

# Add layout enhancements
fig.update_layout(
    title=dict(
        text='Top 10 Colleges with Maximum Selections',
        x=0.5,                      # Center-align the title
        font=dict(size=18, color='darkgreen')  # Use a dark green title for consistency
    ),
    xaxis_tickangle=-45,             # Rotate x-axis labels for better readability
    font=dict(size=14),              # Increase font size
    xaxis=dict(
        title=dict(font=dict(size=16, color='black')),
        tickfont=dict(size=12)
    ),
    yaxis=dict(
        title=dict(font=dict(size=16, color='black'))
    ),
    plot_bgcolor='white',            # Set background to white
    height=600,                      # Increase height for better visuals
    width=900,                       # Increase width
    margin=dict(l=50, r=50, t=100, b=100),  # Adjust margins
    coloraxis_colorbar=dict(
        title='Selections',          # Colorbar title
        titlefont=dict(size=12)
    )
)

# Add gridlines for clarity
fig.update_xaxes(showgrid=False)     # Hide gridlines for x-axis
fig.update_yaxes(showgrid=True, gridwidth=1, gridcolor='lightgrey')

# Show chart
fig.show()
